src/lib.rs

```diff
@@ -160,35 +186,32 @@ impl AsciiChars {
         // ignore unrelated leading bits to find the index of the
         // first related character (if any).
 
-        let matching_bytes: usize;
+        let mut matching_bytes: u64;
 
         unsafe {
-            asm!("pcmpestrm $$0, ($1), $2;"
+            asm!("movlhps $2, $1
+                  pcmpestrm $$0, ($3), $1"
                  : // output operands
                  "={xmm0}"(matching_bytes)
                  : // input operands
-                 "r"(ptr),
                  "x"(self.needle),
-                 "{rdx}"(16u64),
+                 "x"(self.needle_hi),
+                 "r"(ptr),
+                 "{rdx}"(offset + len), // saturates at 16
                  "{rax}"(self.count as u64)
                  : // clobbers
+                 "cc"
                  : // options
             );
         }
 
         // Ignore matches that occurred before our string began
-        let matching_bytes = matching_bytes >> offset;
+        matching_bytes >>= offset;
 
         if matching_bytes != 0 {
-            // Matched somewhere in there, pull out the index
+            // Matched somewhere in there, find the least significant bit
             let index = matching_bytes.trailing_zeros() as usize;
-
-            if index > len {
```

Comment:
It should be ok. It's a bit hard to parse the manual's super generic description of what the instruction does.